File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int binarysearch(int arr[] , int n , int target , bool leftBais){
    int left = 0;
    int right = n-1;
    int i = -1;
    while(left <= right){
        int mid = left + (right - left) / 2;
        if(arr[mid] > target){
            right = mid - 1;
        }
        else if(arr[mid] < target){
            left = mid + 1;
        }
        else{
            i = mid;
            if(leftBais){
                right = mid - 1;
            }
            else{
                left = mid + 1;
            }
        }
    }
    return i;
}

int* searchRange(int* arr, int n, int target, int* returnSize) {
    int* result = (int*)malloc(2 * sizeof(int));
    *returnSize = 2;
    int left = binarysearch(arr , n , target , true);
    int right = binarysearch(arr , n , target , false);
    result[0] = left;
    result[1] = right;
    return result;
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {

    int lowerBound(int[] nums, int x) {
        int low = 0;
        int high = nums.length - 1;
        int ans = nums.length;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (nums[mid] >= x) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return ans;
    }

    int upperBound(int[] nums, int x) {
        int low = 0;
        int high = nums.length - 1;
        int ans = nums.length;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (nums[mid] > x) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return ans;
    }

    public int[] searchRange(int[] nums, int target) {

        int lb = lowerBound(nums, target);

        if (lb == nums.length || nums[lb] != target) {
            return new int[]{-1, -1};
        }

        int ub = upperBound(nums, target);

        return new int[]{lb, ub - 1};
    }
}